### ISSUE TITLE
Understanding the SDPA design on CPU — primitive API, graph dispatch, and intended user path

### ISSUE THREAD

**Issue by sanil-joshi** [oneapi-src/oneDNN#5183]
I  am reading through oneDNN's SDPA-related code to understand the current design. While tracing the dispatch flow on CPU, I noticed a few choices that look deliberate, and I would like to confirm my interpretation before drawing conclusions for my own evaluation.

All references are against upstream `main` .

## Findings

1. **`primitive_kind::sdpa` is library-wide, but CPU returns an empty implementation list**

<img width="595" height="86" alt="Image" src="https://github.com/user-attachments/assets/40b30ac3-f6a6-4586-94d6-20fa47902214" />

<img width="553" height="81" alt="Image" src="https://github.com/user-attachments/assets/139be504-673b-4ff8-966c-8e201d2eb5cf" />
  
There is no src/cpu/cpu_sdpa_list.cpp and no src/cpu/sdpa/ directory.

2. **GPU has full SDPA support (Intel vendor only)**
<img width="558" height="95" alt="Image" src="https://github.com/user-attachments/assets/41cda891-e04f-463c-a3a1-9fa3458866c4" />

3.  **The SDPA primitive surface was deliberately marked internal**
 PR #2549 ("xe: Add internal tests for SDPA", merged 2025-02-25) and its refactor commit b6b9fe9 ("sdpa: refactor internal primitive into   sdpa_test_iface"):

 Deleted src/common/sdpa_internal.{h,hpp}.
 Renamed C entry dnnl_sdpa_primitive_desc_create → sdpa_primitive_desc_create (dropping the public dnnl_ prefix).
 Renamed src/common/sdpa.cpp → src/common/sdpa_test_iface.cpp.
 Moved the C++ wrapper into tests/gtests/internals/sdpa_internal.hpp.
 PR description (quoted verbatim):

 This PR adds tests to directly test the internal micro SDPA primitive. These tests do not access the SDPA primitive through the graph API, instead  the tests are accessing the primitive through exposing a C API symbol for the primitive. ... These tests will allow us to prototype new features to the  SDPA primitive and add tests before the Graph API can expose them through that API.

There is no public dnnl::sdpa class in include/oneapi/dnnl/dnnl.hpp.

4. **The graph backend has a fused single-primitive dispatch path that is gated to GPU**

<img width="575" height="322" alt="Image" src="https://github.com/user-attachments/assets/93daf743-1ce4-4442-a9a9-2948a2be2cb5" />

On CPU enable_ukernel is never set, so sdp_primitive_kernel_t is never attempted. CPU dispatch flows through sdp_decomp_kernel_t (or the generic fallback).

5. **On CPU, the Graph API produces a decomposed execution**
ONEDNN_VERBOSE=all, BF16, batch=32, heads=16, seq=384, head_dim=64:
onednn_verbose,v1,primitive,exec,cpu,matmul,brg_matmul:avx512_core_bf16,undef,src:bf16::blocked:ab::f0 wei:bf16::blocked:ba::f0 dst:f32::blocked:ab::f0,attr-scratchpad:user attr-post-ops:binary_div:bf16:0+binary_add:bf16:3:ab,,384x64:64x384,0.11499
onednn_verbose,v1,primitive,exec,cpu,softmax,jit:avx512_core_bf16,forward_inference,src:f32::blocked:ab::f0 dst:bf16::blocked:ab::f0,attr-scratchpad:user,alg:softmax_accurate_inf_as_zero axis:1,384x384,0.059082
onednn_verbose,v1,primitive,exec,cpu,matmul,brg_matmul:avx512_core_bf16,undef,src:bf16::blocked:ab::f0 wei:bf16::blocked:ab::f0 dst:bf16::blocked:ab::f0,attr-scratchpad:user,,384x384:384x64,0.0849609
onednn_verbose,v1,graph,exec,cpu,100008,sdp,bmm1;scale_div;mask_add;softmax;bmm2,,in0_bf16:0:strided:undef:32x16x384x64:393216s24576s64s1 in1_bf16:1:strided:undef:32x16x384x64:393216s24576s64s1 in2_bf16:4:strided:undef:1:1 in3_bf16:7:strided:undef:32x1x384x384:147456s147456s384s1 in4_bf16:12:strided:undef:32x16x384x64:393216s24576s64s1 out0_bf16:13:strided:undef:32x16x384x64:393216s24576s64s1,fpm:strict,sdp_decomp_kernel_t,dnnl_backend,902.107

**Question**
Could you help me understand the design intent behind the current CPU SDPA arrangement? Specifically:
	• Is the choice to keep sdpa_primitive_desc_create internal (PR #2549) and to gate the graph backend's sdp_primitive_kernel_t to GPU only (Finding 4) a deliberate, long-term design — i.e., users are expected to use the Graph API, which on CPU is intentionally a decomposed path (Finding 5) rather than a single fused primitive call?
	• Or is the current state a transitional one — e.g., the gating in sdp.hpp exists so that when a fused CPU SDPA implementation is added, the dispatch path will activate, and the internal status of the primitive API is pending its eventual stabilization?

I want to make sure I am interpreting the architecture correctly before drawing conclusions. I'm not requesting any feature work — just trying to understand the intent so I can document my findings accurately.

Thanks!

**Comment by TaoLv** [comment]
> the gating in sdp.hpp exists so that when a fused CPU SDPA implementation is added, the dispatch path will activate

@sanil-joshi  Thank you for the question. This is correct. We can dispatch to it once a CPU fused SDPA implementation is added. Currently we have below kernels dispatched based on different problems and architectures:

CPU:
    - larger_partition_kernel_t: sequential primitive calls (matmul w/ post-ops -> softmax -> matmul)
    - sdp_decomp_kernel_t: flash-attention based parallel decomposed primitive calls.
GPU:
    - larger_partition_kernel_t: sequential primitive calls (matmul w/ post-ops -> softmax -> matmul)
    - sdp_primitive_kernel_t: flash-attention based fused SDPA kernel

**Comment by sanil-joshi** [comment]
Thanks @TaoLv  for the information — that clarifies the dispatch ladder.

One more clarification on the user-facing API :
Unlike other operators (matmul, softmax, convolution, etc.), SDPA does not have a public C++ primitive API — there is no `dnnl::sdpa`
class in `include/oneapi/dnnl/dnnl.hpp`, and `sdpa_primitive_desc_create was moved to an internal symbol in PR #2549.

Is there a specific reason for this asymmetry? 

Is the absence of a public primitive API for SDPA an intentional, long-term design choice (e.g., Graph API is the intended user surface for SDPA on both CPU and GPU) ?

**Comment by TaoLv** [comment]
> Is the absence of a public primitive API for SDPA an intentional, long-term design choice (e.g., Graph API is the intended user surface for SDPA on both CPU and GPU) ?

Yes, it's intentional. The purpose is to leverage the flexibility of graph API to cover the wide variants and emerging algorithms of SDPA and other complex fusions, and avoid frequently breaking or bloating the library's API. You can check how the API is integrated in [PyTorch upstream](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp#L180).
